### DISCUSSION TITLE
Tutorial to Run Trilium Server in Termux on Android

### DISCUSSION THREAD

**Discussion by Nriver** [TriliumNext/Trilium#4542] [Show and tell]
## Tutorial to Run Trilium Server in Termux on Android

1. Install [Termux](https://termux.dev/en/) on your Android device.

2. Open Termux and install [PRoot Distro](https://github.com/termux/proot-distro). PRoot is not a virtual machine; you can find more about PRoot Distro [here](https://github.com/termux/proot-distro/blob/master/README.md).
    ```bash
    pkg install proot-distro
    ```

3. Install a base system; here, I use Manjaro. You can find all distros by running `proot-distro list`. I also renamed it for easy typing.
    ```bash
    proot-distro install manjaro-aarch64
    proot-distro rename manjaro-aarch64 manjaro
    ```

    Update: proot-distro has change the name to manjaro by default now. The new command would be this:
    ```bash
    proot-distro install manjaro
    ```

4. Log in to Manjaro
    ```bash
    proot-distro login manjaro
    ```

5. Install Node.js. You can only get the latest version of Node.js by default. So, we need `nvm` to install the specific version of Node.js that can run Trilium; here, I use 18.18.2. (Note: `n` does not work; only `nvm` works.)
    ```bash
    pacman -S nodejs npm base-devel xz
    npm install -g npm node-gyp 
    curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.39.4/install.sh | bash
    nvm install v18.18.2
    nvm alias default v18.18.2
    node -v
    ```

6. Get trilium-server from the [release page](https://github.com/zadam/trilium/releases)
    ```bash
    wget https://github.com/zadam/trilium/releases/download/v0.62.4/trilium-linux-x64-server-0.62.4.tar.xz
    tar -xf trilium-linux-x64-server-0.62.4.tar.xz
    mv trilium-linux-x64-server trilium
    cd trilium
    ```

7. Remove the old dependency folder. As the executables are for `x86`, we do not need them.
    ```bash
    rm -rf node_modules
    ```

8. This is the tricky part; take your time. Reinstall the dependency for `arm`. You may get an error here; you may need to retry the command several times or use a proxy.
    ```bash
    npm install
    npm run switch-server
    ```

9. Start Trilium server
    ```bash
    node src/www
    ```

10. You can access it on your phone's browser with `http://127.0.0.1:8080`.
    Or, you can access it on your computer with `http://your_phone_ip:8080`.

11. For a quick startup of the service, create a `.sh` file in Manjaro
    ```bash
    proot-distro login manjaro
    echo 'cd /root/trilium && node src/www' > /root/run-trilium.sh
    chmod +x /root/run-trilium.sh
    exit
    ```

    Then you can quickly start Trilium in Termux with this single command
    ```bash
    proot-distro login manjaro -- /bin/sh /root/run-trilium.sh
    ```

### How to use a proxy
If you fail to execute `wget`, `npm install`, or `git clone` due to an internet issue, you can use a proxy with `proxychains`. For example:

```bash
pacman -S proxychains

# Change the proxy settings in the last line
# For example: http 127.0.0.1 1080
vim /etc/proxychains.conf

# Execute commands with proxychains
proxychains npm install
proxychains npm run switch-server
```

### How to clear the cache
If `npm install` keeps failing, you can remove the cache and reinstall it.

```bash
rm -rf node_modules
rm -rf ~/.npm/_cacache
```

### Why is it slow?
If you find Trilium is slow, please try to disable power-saving mode on your phone. And give Termux privileges to run in the background or something like that.

**Comment by Nriver** [comment]
Start the server in termux

![IMG_20231228_233436.png](https://github.com/zadam/trilium/assets/6752679/a2663c19-2479-46bf-9995-802caeb0ef7c)

Access the server in your browser

![IMG_20231228_233459.png](https://github.com/zadam/trilium/assets/6752679/9dc76aa4-dc58-4d67-b99c-22e3a1106ba3)

**Reply by Naeem836** [reply]
pleas hellp me how can i downlode termux

**Comment by Ogdestroy404** [comment]
I have a problem!

Listening on port 8080
Sending message to all clients: {"type":"reload-frontend","reason":"source code change"}
Error: Router not found for request GET /assets/v0.62.4/app/setup.js
    at /root/trilium/src/routes/error_handlers.js:18:21
    at Layer.handle [as handle_request] (/root/trilium/node_modules/express/lib/router/layer.js:95:5)
    at trim_prefix (/root/trilium/node_modules/express/lib/router/index.js:328:13)
    at /root/trilium/node_modules/express/lib/router/index.js:286:9
    at Function.process_params (/root/trilium/node_modules/express/lib/router/index.js:346:12)
    at next (/root/trilium/node_modules/express/lib/router/index.js:280:10)
    at Layer.handle [as handle_request] (/root/trilium/node_modules/express/lib/router/layer.js:91:12)
    at trim_prefix (/root/trilium/node_modules/express/lib/router/index.js:328:13)
    at /root/trilium/node_modules/express/lib/router/index.js:286:9
    at Function.process_params (/root/trilium/node_modules/express/lib/router/index.js:346:12) {
  status: 404
}

**Reply by Nriver** [reply]
It's my mistake. Please use `node src/www` to start the service instead of `npm run start-server`. I have already updated the tutorial.

**Comment by Joackk** [comment]
大佬，执行proot被提示了Error: proot-distro should not be executed under PRoot.  我是已经root的手机，这是不能root的用吗？

**Reply by Nriver** [reply]
这个问题已通过重装termux解决。 This is solvded by reinstall existing termux.

**Comment by dnohq** [comment]
感谢大佬，本人小白，记一下自己成功运行的部分
```
pkg update && pkg upgrade
pkg install proot-distro
proot-distro install manjaro
proot-distro login manjaro
pacman -S nodejs npm base-devel xz
npm install -g npm node-gyp
curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.39.4/install.sh | bash
touch ~/.bashrc
echo 'export NVM_DIR="$HOME/.nvm"' >> ~/.bashrc
echo '[ -s "$NVM_DIR/nvm.sh" ] && \. "$NVM_DIR/nvm.sh"' >> ~/.bashrc
echo '[ -s "$NVM_DIR/bash_completion" ] && \. "$NVM_DIR/bash_completion"' >> ~/.bashrc
source ~/.bashrc
nvm install v18.18.2
nvm alias default v18.18.2
node -v
pacman -S wget
wget https://github.com/zadam/trilium/releases/download/v0.62.4/trilium-linux-x64-server-0.62.4.tar.xz
tar -xf trilium-linux-x64-server-0.62.4.tar.xz
mv trilium-linux-x64-server trilium
cd trilium
rm -rf node_modules
npm install
npm run switch-server
node src/www
```
然后启动用
```
proot-distro login manjaro
cd trilium
npm rebuild better-sqlite3
node src/www
```

**Comment by totalscout** [comment]
Damm, this is both grazy and amazing!
